himbaechel/uarch/gatemate/bitstream.cc

```diff
@@ -0,0 +1,258 @@
+            case id_CC_IBUF.index:
+            case id_CC_TOBUF.index:
+            case id_CC_OBUF.index:
+            case id_CC_IOBUF.index:
+            case id_CC_LVDS_IBUF.index:
+            case id_CC_LVDS_TOBUF.index:
+            case id_CC_LVDS_OBUF.index:
+            case id_CC_LVDS_IOBUF.index:
+                for (auto &p : params) {
+                    cc.tiles[loc].add_word(stringf("GPIO.%s", p.first.c_str(ctx)), p.second.as_bits());
+                }
+                break;
+            case id_CPE_HALF_U.index:
+            case id_CPE_HALF_L.index: {
+                int id = tile_extra_data(cell.second.get()->bel.tile)->prim_id;
+                for (auto &p : params) {
+                    cc.tiles[loc].add_word(stringf("CPE%d.%s", id, p.first.c_str(ctx)), p.second.as_bits());
+                }
+            } break;
+            case id_BUFG.index: {
+                Loc l = ctx->getBelLocation(cell.second->bel);
+                cc.configs[0].add_word(stringf("GLBOUT.GLB%d_EN", l.z), int_to_bitvector(1, 1));
+            } break;
+            case id_PLL.index: {
+                Loc l = ctx->getBelLocation(cell.second->bel);
+                for (auto &p : params) {
+                    cc.configs[0].add_word(stringf("PLL%d.%s", l.z - 4, p.first.c_str(ctx)), p.second.as_bits());
+                }
+            } break;
+            case id_RAM.index: {
+                CfgLoc loc = getRAMConfigLoc(cell.second.get()->bel.tile);
+                auto &bram = cc.brams[loc];
+                for (auto &p : params) {
+                    std::string name = p.first.c_str(ctx);
+                    if (boost::starts_with(name, "RAM_cfg"))
+                        bram.add_word(name, p.second.as_bits());
+                }
+
+                bool is_fifo = params.count(id_RAM_cfg_fifo_sync_enable) | params.count(id_RAM_cfg_fifo_async_enable);
+                if (!is_fifo) {
+                    auto &bram_data = cc.bram_data[loc];
+                    bram_data = std::vector<uint8_t>(5120);
+                    for (int i = 0; i < 128; i++) {
+                        for (int j = 0; j < 40; j++) {
+                            bram_data[i * 40 + j] = extract_bits(params, ctx->idf("INIT_%02X", i), j * 8, 8);
+                        }
+                    }
+                }
+            } break;
+            case id_USR_RSTN.index:
+            case id_CFG_CTRL.index:
+                break;
+            default:
+                log_error("Unhandled cell %s of type %s\n", cell.second.get()->name.c_str(ctx),
+                          cell.second->type.c_str(ctx));
+            }
+        }
+
+        for (auto &net : ctx->nets) {
+            NetInfo *ni = net.second.get();
+            if (ni->wires.empty())
+                continue;
+            std::set<std::string> nets;
+            for (auto &w : ni->wires) {
+                if (w.second.pip != PipId()) {
+                    PipId pip = w.second.pip;
```

Comment:
this big chunk could maybe be factored into a function of its own to stop `write_bitstream` becoming too big (lessons learnt personally from ecp5)

Reply:
refactored this now is separate function